### PR TITLE
[Remote Store] Optimize segments metadata upload

### PR DESCRIPTION
**Pull request by ashking94**
### Description
[Describe what this change achieves]

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java [hunk 1]
```diff
@@ -230,23 +234,25 @@ private synchronized void syncSegments(boolean isRetry) {
...
-                                storeDirectory,
-                                indexShard.getOperationPrimaryTerm(),
-                                segmentInfos.getGeneration()
-                            );
-                            clearStaleFilesFromLocalSegmentChecksumMap(localSegmentsPostRefresh);
-                            onSuccessfulSegmentsSync(refreshTimeMs, refreshSeqNo);
-                            indexShard.getEngine().translogManager().setMinSeqNoToKeep(lastRefreshedCheckpoint + 1);
-                            checkpointPublisher.publish(indexShard, checkpoint);
+                        List<String> segmentFilesToUpload = localSegmentsPostRefresh.stream()
+                            .filter(f -> skipUpload(f) == false)
```
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java:238]:
```suggestion
                            .filter(f -> !skipUpload(f))
```

File: server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java [hunk 2]
```diff
@@ -230,23 +234,25 @@ private synchronized void syncSegments(boolean isRetry) {
...
+                            .collect(Collectors.toList());
+
+                        if (segmentFilesToUpload.isEmpty() || uploadNewSegments(segmentFilesToUpload)) {
+                            if (segmentFilesInLatestMetadata.equals(localSegmentsPostRefresh) == false) {
```
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java:242]:
```suggestion
                            if (!segmentFilesInLatestMetadata.equals(localSegmentsPostRefresh)) {
```

File: server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java [hunk 3]
```diff
@@ -283,11 +289,7 @@ private synchronized void syncSegments(boolean isRetry) {
      * @param localSegmentsPostRefresh list of segment files present post refresh
      */
     private void clearStaleFilesFromLocalSegmentChecksumMap(Collection<String> localSegmentsPostRefresh) {
-        localSegmentChecksumMap.keySet()
-            .stream()
-            .filter(file -> !localSegmentsPostRefresh.contains(file))
-            .collect(Collectors.toSet())
-            .forEach(localSegmentChecksumMap::remove);
+        localSegmentChecksumMap.entrySet().removeIf(entry -> localSegmentsPostRefresh.contains(entry.getKey()) == false);
```
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java:292]:
```suggestion
        localSegmentChecksumMap.entrySet().removeIf(entry -> !localSegmentsPostRefresh.contains(entry.getKey()));
```

File: server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java [hunk 4]
```diff
@@ -365,6 +371,25 @@ String uploadSegmentInfosSnapshot(String latestSegmentsNFilename, SegmentInfos s
         return segmentInfoSnapshotFilename;
     }
 
+    /**
+     * Whether to upload a file or not depending on whether file is in excluded list or has been already uploaded.
+     *
+     * @param file that needs to be uploaded.
```
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java:377]:
```suggestion
     * @param file name of the file that needs to be uploaded.
```

**Comment by ashking94** [comment]
Thanks @dbwiddis , but this was only a draft PR raised for comparing 2 different solutions of the same problem.

**Pull request closed without merging**